src/main/java/duke/commands/EditDateCommand.java

```diff
@@ -0,0 +1,168 @@
+                                        " changed from...");
+                                DukeUI.commandWrap(DukeUI.NOTE_TIME.format(this.oldDate) + " ...and... " +
+                                        DukeUI.NOTE_TIME.format(oldEnd), 66);
+                                System.out.println("    to...");
+                                DukeUI.commandWrap(DukeUI.NOTE_TIME.format(this.newDate) + " ...and... " +
+                                        DukeUI.NOTE_TIME.format(newEnd), 66);
+                                DukeUI.autoSaveConfirmation(new SaveCommand().autoSave(dukeNotes, dukeStorage));
+                                DukeUI.suggestListNotes();
+                            }
+                        } else {
+                            throw new DateException(this.newDate, "NoStart");
+                        }
+                        break;
+                    case "end":
+                        if(dukeNotes.getNotes().get(i) instanceof Event) {
+                            if (dukeNotes.getNotes().get(i).getIsDone()) {
+                                DukeUI.printDivider();
+                                System.out.println("    The event had already concluded.");
+                                System.out.println("    The date shouldn't be edited anymore.");
+                            } else {
+                                this.oldDate = ((Event) dukeNotes.getNotes().get(i)).getEndDate();
+                                Date oldStart = ((Event) dukeNotes.getNotes().get(i)).getStartDate();
+
+                                ArrayList<Note> notesLess_i = new ArrayList<>(dukeNotes.getNotes());
+                                notesLess_i.remove(i);
+                                NewNoteCommand.checkForClashes(notesLess_i, oldStart, newDate);
+
+                                ((Event) dukeNotes.getNotes().get(i)).setEndDate(this.newDate);
+
+                                DukeUI.printDivider();
+                                System.out.println("    End Date of Note #" + this.targetNote + " changed from...");
+                                DukeUI.commandWrap(DukeUI.NOTE_TIME.format(this.oldDate), 66);
+                                System.out.println("    to...");
+                                DukeUI.commandWrap(DukeUI.NOTE_TIME.format(this.newDate), 66);
+                                DukeUI.autoSaveConfirmation(new SaveCommand().autoSave(dukeNotes, dukeStorage));
+                                DukeUI.suggestListNotes();
+                            }
+                        } else {
+                            throw new DateException(this.newDate, "NoEnd");
+                        }
+                        break;
+                    default:
+                        throw new CommandException("The type of date you are trying to edit does not exist.");
+                }
+                DukeUI.printDivider();
+                break;
+            }
```

Comment:
Consider extracting code of the each `case` block as a separate method. Within each of those methods, you can use _guard clauses_ to reduce nesting.

Reply:
Thank you Prof, the improvements have been made.